{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "# tensorflowがgpuを使っているか確認する\n",
    "import tensorflow as tf\n",
    "print(tf.test.is_built_with_cuda())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"tuneStatus\">\n",
       "  <div style=\"display: flex;flex-direction: row\">\n",
       "    <div style=\"display: flex;flex-direction: column;\">\n",
       "      <h3>Tune Status</h3>\n",
       "      <table>\n",
       "<tbody>\n",
       "<tr><td>Current time:</td><td>2024-02-22 01:08:34</td></tr>\n",
       "<tr><td>Running for: </td><td>00:04:26.39        </td></tr>\n",
       "<tr><td>Memory:      </td><td>62.8/63.2 GiB      </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "    </div>\n",
       "    <div class=\"vDivider\"></div>\n",
       "    <div class=\"systemInfo\">\n",
       "      <h3>System Info</h3>\n",
       "      Using FIFO scheduling algorithm.<br>Logical resource usage: 11.0/24 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:G)\n",
       "    </div>\n",
       "    <div class=\"vDivider\"></div>\n",
       "<div class=\"messages\">\n",
       "  <h3>Messages</h3>\n",
       "  : ***LOW MEMORY*** less than 10% of the memory on this node is available for use. This can cause unexpected crashes. Consider reducing the memory used by your application or reducing the Ray object store size by setting `object_store_memory` when calling `ray.init`.\n",
       "  \n",
       "  \n",
       "</div>\n",
       "<style>\n",
       ".messages {\n",
       "  color: var(--jp-ui-font-color1);\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  padding-left: 1em;\n",
       "  overflow-y: auto;\n",
       "}\n",
       ".messages h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".vDivider {\n",
       "  border-left-width: var(--jp-border-width);\n",
       "  border-left-color: var(--jp-border-color0);\n",
       "  border-left-style: solid;\n",
       "  margin: 0.5em 1em 0.5em 1em;\n",
       "}\n",
       "</style>\n",
       "\n",
       "  </div>\n",
       "  <div class=\"hDivider\"></div>\n",
       "  <div class=\"trialStatus\">\n",
       "    <h3>Trial Status</h3>\n",
       "    <table>\n",
       "<thead>\n",
       "<tr><th>Trial name                  </th><th>status  </th><th>loc  </th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>DreamerV3_my_env_d84dc_00000</td><td>PENDING </td><td>     </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "  </div>\n",
       "</div>\n",
       "<style>\n",
       ".tuneStatus {\n",
       "  color: var(--jp-ui-font-color1);\n",
       "}\n",
       ".tuneStatus .systemInfo {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus td {\n",
       "  white-space: nowrap;\n",
       "}\n",
       ".tuneStatus .trialStatus {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".tuneStatus .hDivider {\n",
       "  border-bottom-width: var(--jp-border-width);\n",
       "  border-bottom-color: var(--jp-border-color0);\n",
       "  border-bottom-style: solid;\n",
       "}\n",
       ".tuneStatus .vDivider {\n",
       "  border-left-width: var(--jp-border-width);\n",
       "  border-left-color: var(--jp-border-color0);\n",
       "  border-left-style: solid;\n",
       "  margin: 0.5em 1em 0.5em 1em;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-02-22 01:04:08,599\tINFO wandb.py:307 -- Already logged into W&B.\n",
      "\u001b[36m(pid=13244)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\utils\\passive_env_checker.py:135: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method was expecting numpy array dtype to be float32, actual type: float64\u001b[0m\n",
      "\u001b[36m(pid=13244)\u001b[0m   logger.warn(\n",
      "\u001b[36m(pid=13244)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\utils\\passive_env_checker.py:159: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method is not within the observation space.\u001b[0m\n",
      "\u001b[36m(pid=13244)\u001b[0m   logger.warn(f\"{pre} is not within the observation space.\")\n",
      "\u001b[36m(pid=44076)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\utils\\passive_env_checker.py:135: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method was expecting numpy array dtype to be float32, actual type: float64\u001b[0m\n",
      "\u001b[36m(pid=44076)\u001b[0m   logger.warn(\n",
      "\u001b[36m(pid=44076)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\utils\\passive_env_checker.py:159: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method is not within the observation space.\u001b[0m\n",
      "\u001b[36m(pid=44076)\u001b[0m   logger.warn(f\"{pre} is not within the observation space.\")\n",
      "\u001b[36m(pid=12812)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\utils\\passive_env_checker.py:135: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method was expecting numpy array dtype to be float32, actual type: float64\u001b[0m\n",
      "\u001b[36m(pid=12812)\u001b[0m   logger.warn(\n",
      "\u001b[36m(pid=12812)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\utils\\passive_env_checker.py:159: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method is not within the observation space.\u001b[0m\n",
      "\u001b[36m(pid=12812)\u001b[0m   logger.warn(f\"{pre} is not within the observation space.\")\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44076)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\vector\\__init__.py:53: UserWarning: \u001b[33mWARN: `gymnasium.vector.make(...)` is deprecated and will be replaced by `gymnasium.make_vec(...)` in v1.0\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44076)\u001b[0m   gym.logger.warn(\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m *** SIGABRT received at time=1708531532 ***\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFADC31F601  (unknown)  abort\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFA573DEB96  (unknown)  PyInit__raylet\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFADC31ED32  (unknown)  terminate\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF72E8E1C6F  (unknown)  OPENSSL_Applink\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFADCA2CF5C  (unknown)  UnhandledExceptionFilter\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFADEFD8ABD  (unknown)  memcpy\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFADEFBF197  (unknown)  _C_specific_handler\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFADEFD441F  (unknown)  _chkstk\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFADEF4E466  (unknown)  RtlFindCharInUnicodeString\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFADEF84475  (unknown)  RtlRaiseException\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFADC935B0C  (unknown)  RaiseException\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FFAD5C46220  (unknown)  CxxThrowException\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9C7A168B3  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9C7A154E1  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A4A55375  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A4EEDFC4  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9C7357924  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9C7354B9B  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9C7353A0A  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9C73547DB  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9C7664CD8  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9C7668E7A  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A6D26B6C  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A6D0993C  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A6D439DB  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A6D07964  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A6CFE76F  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A6CFD3B7  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A6D0B2E3  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A6D0065F  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A6CFF20F  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m     @   00007FF9A61FFE5C  (unknown)  (unknown)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m Fatal Python error: Aborted\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m \n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m Stack (most recent call first):\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 102 in convert_to_eager_tensor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 304 in _constant_eager_impl\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 279 in _constant_impl\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 267 in constant\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\tensor_conversion_registry.py\", line 48 in _default_conversion_function\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 1638 in convert_to_tensor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\profiler\\trace.py\", line 183 in wrapped\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 1817 in _init_from_args\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 1659 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 268 in __call__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150 in error_handler\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variable_scope.py\", line 2755 in default_variable_creator_v2\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 240 in <lambda>\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 247 in _variable_v2_call\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 266 in __call__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150 in error_handler\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\engine\\training.py\", line 334 in _init_batch_counters\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\trackable\\base.py\", line 205 in _method_wrapper\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\engine\\training.py\", line 319 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65 in error_handler\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\trackable\\base.py\", line 205 in _method_wrapper\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\tf\\tf_rl_module.py\", line 17 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 377 in new_init\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 377 in new_init\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 104 in build\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\marl_module.py\", line 77 in setup\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 369 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\marl_module.py\", line 70 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 377 in new_init\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\marl_module.py\", line 491 in build\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\algorithms\\dreamerv3\\utils\\env_runner.py\", line 146 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\util\\tracing\\tracing_helper.py\", line 467 in _resume_span\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\function_manager.py\", line 726 in actor_method_executor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\worker.py\", line 847 in main_loop\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\workers\\default_worker.py\", line 282 in <module>\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m \n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m Extension modules: msgpack._cmsgpack, psutil._psutil_windows, setproctitle, yaml._yaml, _brotli, charset_normalizer.md, ray._raylet, mkl._mklinit, mkl._py_mkl_service, numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, h5py._errors, h5py.defs, h5py._objects, h5py.h5, h5py.utils, h5py.h5t, h5py.h5s, h5py.h5ac, h5py.h5p, h5py.h5r, h5py._proxy, h5py._conv, h5py.h5z, h5py.h5a, h5py.h5d, h5py.h5ds, h5py.h5g, h5py.h5i, h5py.h5f, h5py.h5fd, h5py.h5pl, h5py.h5o, h5py.h5l, h5py._selector, scipy._lib._ccallback_c, scipy.sparse._sparsetools, _csparsetools, scipy.sparse._csparsetools, scipy.linalg._fblas, scipy.linalg._flapack, scipy.linalg.cython_lapack, scipy.linalg._cythonized_array_utils, scipy.linalg._solve_toeplitz, scipy.linalg._flinalg, scipy.linalg._decomp_lu_cython, scipy.linalg._matfuncs_sqrtm_triu, scipy.linalg.cython_blas, scipy.linalg._matfuncs_expm, scipy.linalg._decomp_update, scipy.sparse.linalg._dsolve._superlu, scipy.sparse.linalg._eigen.arpack._arpack, scipy.sparse.csgraph._tools, scipy.sparse.csgraph._shortest_path, scipy.sparse.csgraph._traversal, scipy.sparse.csgraph._min_spanning_tree, scipy.sparse.csgraph._flow, scipy.sparse.csgraph._matching, scipy.sparse.csgraph._reordering, lz4._version, lz4.frame._frame, PIL._imaging, pyarrow.lib, pyarrow._hdfsio, pandas._libs.tslibs.ccalendar, pandas._libs.tslibs.np_datetime, pandas._libs.tslibs.dtypes, pandas._libs.tslibs.base, pandas._libs.tslibs.nattype, pandas._libs.tslibs.timezones, pandas._libs.tslibs.fields, pandas._libs.tslibs.timedeltas, pandas._libs.tslibs.tzconversion, pandas._libs.tslibs.timestamps, pandas._libs.properties, pandas._libs.tslibs.offsets, pandas._libs.tslibs.strptime, pandas._libs.tslibs.parsing, pandas._libs.tslibs.conversion, pandas._libs.tslibs.period, pandas._libs.tslibs.vectorized, pandas._libs.ops_dispatch, pandas._libs.missing, pandas._libs.hashtable, pandas._libs.algos, pandas._libs.interval, pandas._libs.lib, pandas._libs.ops, pandas._libs.hashing, pandas._libs.arrays, pandas._libs.tslib, pandas._libs.sparse, pandas._libs.internals, pandas._libs.indexing, pandas._libs.index, pandas._libs.writers, pandas._libs.join, pandas._libs.window.aggregations, pandas._libs.window.indexers, pandas._libs.reshape, pandas._libs.groupby, pandas._libs.json, pandas._libs.parsers, pandas._libs.testing, scipy.ndimage._nd_image, scipy.special._ufuncs_cxx, scipy.special._ufuncs, scipy.special._specfun, scipy.special._comb, scipy.special._ellip_harm_2, _ni_label, scipy.ndimage._ni_label, torch._C, torch._C._fft, torch._C._linalg, torch._C._nested, torch._C._nn, torch._C._sparse, torch._C._special, pyarrow._fs, pyarrow._hdfs, pyarrow._s3fs, pyarrow._parquet, scipy.signal._sigtools, scipy._lib._uarray._uarray, scipy.signal._max_len_seq_inner, scipy.signal._upfirdn_apply, scipy.signal._spline, scipy.spatial._ckdtree, scipy._lib.messagestream, scipy.spatial._qhull, scipy.spatial._voronoi, scipy.spatial._distance_wrap, scipy.spatial._hausdorff, scipy.spatial.transform._rotation, scipy.integrate._odepack, scipy.integrate._quadpack, scipy.integrate._vode, scipy.integrate._dop, scipy.integrate._lsoda, scipy.optimize._minpack2, scipy.optimize._group_columns, scipy.optimize._trlib._trlib, scipy.optimize._lbfgsb, _moduleTNC, scipy.optimize._moduleTNC, scipy.optimize._cobyla, scipy.optimize._slsqp, scipy.optimize._minpack, scipy.optimize._lsq.givens_elimination, scipy.optimize._zeros, scipy.optimize._highs.cython.src._highs_wrapper, scipy.optimize._highs._highs_wrapper, scipy.optimize._highs.cython.src._highs_constants, scipy.optimize._highs._highs_constants, scipy.linalg._interpolative, scipy.optimize._bglu_dense, scipy.optimize._lsap, scipy.optimize._direct, scipy.interpolate._fitpack, scipy.interpolate.dfitpack, scipy.interpolate._bspl, scipy.interpolate._ppoly, scipy.interpolate.interpnd, scipy.interpolate._rbfinterp_pythran, scipy.interpolate._rgi_cython, scipy.signal._sosfilt, scipy.signal._spectral, scipy.special.cython_special, scipy.stats._stats, scipy.stats.beta_ufunc, scipy.stats._boost.beta_ufunc, scipy.stats.binom_ufunc, scipy.stats._boost.binom_ufunc, scipy.stats.nbinom_ufunc, scipy.stats._boost.nbinom_ufunc, scipy.stats.hypergeom_ufunc, scipy.stats._boost.hypergeom_ufunc, scipy.stats.ncf_ufunc, scipy.stats._boost.ncf_ufunc, scipy.stats.ncx2_ufunc, scipy.stats._boost.ncx2_ufunc, scipy.stats.nct_ufunc, scipy.stats._boost.nct_ufunc, scipy.stats.skewnorm_ufunc, scipy.stats._boost.skewnorm_ufunc, scipy.stats.invgauss_ufunc, scipy.stats._boost.invgauss_ufunc, scipy.stats._biasedurn, scipy.stats._levy_stable.levyst, scipy.stats._stats_pythran, scipy.stats._ansari_swilk_statistics, scipy.stats._sobol, scipy.stats._qmc_cy, scipy.stats._mvn, scipy.stats._rcont.rcont, scipy.stats._unuran.unuran_wrapper, scipy.signal._peak_finding_utils, scipy.fftpack.convolve, mkl_fft._pydfti, scipy.io.matlab._mio_utils, scipy.io.matlab._streams, scipy.io.matlab._mio5_utils, matplotlib._c_internal_utils, matplotlib._path, kiwisolver._cext, matplotlib._image (total: 218)\n",
      "\u001b[36m(pid=28612)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\utils\\passive_env_checker.py:135: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method was expecting numpy array dtype to be float32, actual type: float64\u001b[0m\u001b[32m [repeated 8x across cluster] (Ray deduplicates logs by default. Set RAY_DEDUP_LOGS=0 to disable log deduplication, or see https://docs.ray.io/en/master/ray-observability/ray-logging.html#log-deduplication for more options.)\u001b[0m\n",
      "\u001b[36m(pid=28612)\u001b[0m   logger.warn(\u001b[32m [repeated 8x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=28612)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\utils\\passive_env_checker.py:159: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method is not within the observation space.\u001b[0m\u001b[32m [repeated 8x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=28612)\u001b[0m   logger.warn(f\"{pre} is not within the observation space.\")\u001b[32m [repeated 8x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=28612)\u001b[0m c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\gymnasium\\vector\\__init__.py:53: UserWarning: \u001b[33mWARN: `gymnasium.vector.make(...)` is deprecated and will be replaced by `gymnasium.make_vec(...)` in v1.0\u001b[0m\u001b[32m [repeated 9x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=28612)\u001b[0m   gym.logger.warn(\u001b[32m [repeated 9x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m Stack (most recent call first):\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 102 in convert_to_eager_tensor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 304 in _constant_eager_impl\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 279 in _constant_impl\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 267 in constant\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\tensor_conversion_registry.py\", line 48 in _default_conversion_function\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 1638 in convert_to_tensor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\profiler\\trace.py\", line 183 in wrapped\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 1817 in _init_from_args\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 1659 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 268 in __call__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150 in error_handler\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variable_scope.py\", line 2755 in default_variable_creator_v2\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 240 in <lambda>\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 247 in _variable_v2_call\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 266 in __call__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\util\\traceback_utils.py\", line 150 in error_handler\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\engine\\training.py\", line 334 in _init_batch_counters\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\trackable\\base.py\", line 205 in _method_wrapper\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\engine\\training.py\", line 319 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65 in error_handler\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\trackable\\base.py\", line 205 in _method_wrapper\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\tf\\tf_rl_module.py\", line 17 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 377 in new_init\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 377 in new_init\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 104 in build\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\marl_module.py\", line 77 in setup\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 369 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\marl_module.py\", line 70 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 377 in new_init\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\marl_module.py\", line 491 in build\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\algorithms\\dreamerv3\\utils\\env_runner.py\", line 146 in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\util\\tracing\\tracing_helper.py\", line 467 in _resume_span\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\function_manager.py\", line 726 in actor_method_executor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\worker.py\", line 847 in main_loop\n",
      "\u001b[36m(DreamerV3EnvRunner pid=21448)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\workers\\default_worker.py\", line 282 in <module>\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m *** SIGABRT received at time=1708531579 ***\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFADC31F601  (unknown)  abort\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFA573DEB96  (unknown)  PyInit__raylet\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFADC31ED32  (unknown)  terminate\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FF72E8E1C6F  (unknown)  OPENSSL_Applink\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFADCA2CF5C  (unknown)  UnhandledExceptionFilter\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFADEFD8ABD  (unknown)  memcpy\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFADEFBF197  (unknown)  _C_specific_handler\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFADEFD441F  (unknown)  _chkstk\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFADEF4E466  (unknown)  RtlFindCharInUnicodeString\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFADEF84475  (unknown)  RtlRaiseException\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFADC935B0C  (unknown)  RaiseException\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FFAD5C46220  (unknown)  CxxThrowException\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m     @   00007FF9A61FFE5C  (unknown)  (unknown)\u001b[32m [repeated 20x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m Fatal Python error: Aborted\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m \u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m Stack (most recent call first):\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 102 in convert_to_eager_tensor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 304 in _constant_eager_impl\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 279 in _constant_impl\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 267 in constant\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\tensor_conversion_registry.py\", line 48 in _default_conversion_function\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 1638 in convert_to_tensor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\profiler\\trace.py\", line 183 in wrapped\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 1817 in _init_from_args\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\algorithms\\dreamerv3\\utils\\env_runner.py\", line 146 in __init__\u001b[32m [repeated 6x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 266 in __call__\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65 in error_handler\u001b[32m [repeated 3x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 247 in _variable_v2_call\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 240 in <lambda>\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\engine\\training.py\", line 334 in _init_batch_counters\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\trackable\\base.py\", line 205 in _method_wrapper\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 377 in new_init\u001b[32m [repeated 3x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\marl_module.py\", line 491 in build\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\marl_module.py\", line 77 in setup\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\util\\tracing\\tracing_helper.py\", line 467 in _resume_span\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\function_manager.py\", line 726 in actor_method_executor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\worker.py\", line 847 in main_loop\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\workers\\default_worker.py\", line 282 in <module>\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m Extension modules: msgpack._cmsgpack, psutil._psutil_windows, setproctitle, yaml._yaml, _brotli, charset_normalizer.md, ray._raylet, mkl._mklinit, mkl._py_mkl_service, numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, h5py._errors, h5py.defs, h5py._objects, h5py.h5, h5py.utils, h5py.h5t, h5py.h5s, h5py.h5ac, h5py.h5p, h5py.h5r, h5py._proxy, h5py._conv, h5py.h5z, h5py.h5a, h5py.h5d, h5py.h5ds, h5py.h5g, h5py.h5i, h5py.h5f, h5py.h5fd, h5py.h5pl, h5py.h5o, h5py.h5l, h5py._selector, scipy._lib._ccallback_c, scipy.sparse._sparsetools, _csparsetools, scipy.sparse._csparsetools, scipy.linalg._fblas, scipy.linalg._flapack, scipy.linalg.cython_lapack, scipy.linalg._cythonized_array_utils, scipy.linalg._solve_toeplitz, scipy.linalg._flinalg, scipy.linalg._decomp_lu_cython, scipy.linalg._matfuncs_sqrtm_triu, scipy.linalg.cython_blas, scipy.linalg._matfuncs_expm, scipy.linalg._decomp_update, scipy.sparse.linalg._dsolve._superlu, scipy.sparse.linalg._eigen.arpack._arpack, scipy.sparse.csgraph._tools, scipy.sparse.csgraph._shortest_path, scipy.sparse.csgraph._traversal, scipy.sparse.csgraph._min_spanning_tree, scipy.sparse.csgraph._flow, scipy.sparse.csgraph._matching, scipy.sparse.csgraph._reordering, lz4._version, lz4.frame._frame, PIL._imaging, pyarrow.lib, pyarrow._hdfsio, pandas._libs.tslibs.ccalendar, pandas._libs.tslibs.np_datetime, pandas._libs.tslibs.dtypes, pandas._libs.tslibs.base, pandas._libs.tslibs.nattype, pandas._libs.tslibs.timezones, pandas._libs.tslibs.fields, pandas._libs.tslibs.timedeltas, pandas._libs.tslibs.tzconversion, pandas._libs.tslibs.timestamps, pandas._libs.properties, pandas._libs.tslibs.offsets, pandas._libs.tslibs.strptime, pandas._libs.tslibs.parsing, pandas._libs.tslibs.conversion, pandas._libs.tslibs.period, pandas._libs.tslibs.vectorized, pandas._libs.ops_dispatch, pandas._libs.missing, pandas._libs.hashtable, pandas._libs.algos, pandas._libs.interval, pandas._libs.lib, pandas._libs.ops, pandas._libs.hashing, pandas._libs.arrays, pandas._libs.tslib, pandas._libs.sparse, pandas._libs.internals, pandas._libs.indexing, pandas._libs.index, pandas._libs.writers, pandas._libs.join, pandas._libs.window.aggregations, pandas._libs.window.indexers, pandas._libs.reshape, pandas._libs.groupby, pandas._libs.json, pandas._libs.parsers, pandas._libs.testing, scipy.ndimage._nd_image, scipy.special._ufuncs_cxx, scipy.special._ufuncs, scipy.special._specfun, scipy.special._comb, scipy.special._ellip_harm_2, _ni_label, scipy.ndimage._ni_label, torch._C, torch._C._fft, torch._C._linalg, torch._C._nested, torch._C._nn, torch._C._sparse, torch._C._special, pyarrow._fs, pyarrow._hdfs, pyarrow._s3fs, pyarrow._parquet, scipy.signal._sigtools, scipy._lib._uarray._uarray, scipy.signal._max_len_seq_inner, scipy.signal._upfirdn_apply, scipy.signal._spline, scipy.spatial._ckdtree, scipy._lib.messagestream, scipy.spatial._qhull, scipy.spatial._voronoi, scipy.spatial._distance_wrap, scipy.spatial._hausdorff, scipy.spatial.transform._rotation, scipy.integrate._odepack, scipy.integrate._quadpack, scipy.integrate._vode, scipy.integrate._dop, scipy.integrate._lsoda, scipy.optimize._minpack2, scipy.optimize._group_columns, scipy.optimize._trlib._trlib, scipy.optimize._lbfgsb, _moduleTNC, scipy.optimize._moduleTNC, scipy.optimize._cobyla, scipy.optimize._slsqp, scipy.optimize._minpack, scipy.optimize._lsq.givens_elimination, scipy.optimize._zeros, scipy.optimize._highs.cython.src._highs_wrapper, scipy.optimize._highs._highs_wrapper, scipy.optimize._highs.cython.src._highs_constants, scipy.optimize._highs._highs_constants, scipy.linalg._interpolative, scipy.optimize._bglu_dense, scipy.optimize._lsap, scipy.optimize._direct, scipy.interpolate._fitpack, scipy.interpolate.dfitpack, scipy.interpolate._bspl, scipy.interpolate._ppoly, scipy.interpolate.interpnd, scipy.interpolate._rbfinterp_pythran, scipy.interpolate._rgi_cython, scipy.signal._sosfilt, scipy.signal._spectral, scipy.special.cython_special, scipy.stats._stats, scipy.stats.beta_ufunc, scipy.stats._boost.beta_ufunc, scipy.stats.binom_ufunc, scipy.stats._boost.binom_ufunc, scipy.stats.nbinom_ufunc, scipy.stats._boost.nbinom_ufunc, scipy.stats.hypergeom_ufunc, scipy.stats._boost.hypergeom_ufunc, scipy.stats.ncf_ufunc, scipy.stats._boost.ncf_ufunc, scipy.stats.ncx2_ufunc, scipy.stats._boost.ncx2_ufunc, scipy.stats.nct_ufunc, scipy.stats._boost.nct_ufunc, scipy.stats.skewnorm_ufunc, scipy.stats._boost.skewnorm_ufunc, scipy.stats.invgauss_ufunc, scipy.stats._boost.invgauss_ufunc, scipy.stats._biasedurn, scipy.stats._levy_stable.levyst, scipy.stats._stats_pythran, scipy.stats._ansari_swilk_statistics, scipy.stats._sobol, scipy.stats._qmc_cy, scipy.stats._mvn, scipy.stats._rcont.rcont, scipy.stats._unuran.unuran_wrapper, scipy.signal._peak_finding_utils, scipy.fftpack.convolve, mkl_fft._pydfti, scipy.io.matlab._mio_utils, scipy.io.matlab._streams, scipy.io.matlab._mio5_utils, matplotlib._c_internal_utils, matplotlib._path, kiwisolver._cext, matplotlib._image (total: 218)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m Stack (most recent call first):\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 102 in convert_to_eager_tensor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 304 in _constant_eager_impl\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 279 in _constant_impl\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\constant_op.py\", line 267 in constant\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\tensor_conversion_registry.py\", line 48 in _default_conversion_function\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\framework\\ops.py\", line 1638 in convert_to_tensor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\profiler\\trace.py\", line 183 in wrapped\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\resource_variable_ops.py\", line 1817 in _init_from_args\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 240 in <lambda>\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\engine\\training.py\", line 334 in _init_batch_counters\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\marl_module.py\", line 77 in setup\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\util\\tracing\\tracing_helper.py\", line 467 in _resume_span\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\function_manager.py\", line 726 in actor_method_executor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\worker.py\", line 847 in main_loop\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\workers\\default_worker.py\", line 282 in <module>\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m Exception raised in creation task: The actor died because of an error raised in its creation task, \u001b[36mray::DreamerV3EnvRunner.__init__()\u001b[39m (pid=12812, ip=127.0.0.1, actor_id=6479705746360d7763b7ed8801000000, repr=<ray.rllib.algorithms.dreamerv3.utils.env_runner.DreamerV3EnvRunner object at 0x0000027BF58C1E10>)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"python\\ray\\_raylet.pyx\", line 1813, in ray._raylet.execute_task\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"python\\ray\\_raylet.pyx\", line 1754, in ray._raylet.execute_task.function_executor\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     return method(__ray_actor, *args, **kwargs)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     return method(self, *_args, **_kwargs)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     self.module = module_spec.build()[DEFAULT_POLICY_ID]\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     module = self.marl_module_class(module_config)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     previous_init(self, *args, **kwargs)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     super().__init__(config or MultiAgentRLModuleConfig())\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     self.setup()\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     self._rl_modules[module_id] = module_spec.build()\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     module = self.module_class(module_config)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     previous_init(self, *args, **kwargs)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     previous_init(self, *args, **kwargs)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     RLModule.__init__(self, *args, **kwargs)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     self.setup()\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     catalog = self.config.get_catalog()\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 189, in get_catalog\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     return self.catalog_class(\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     super().__init__(\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     self._determine_components_hook()\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\models\\catalog.py\", line 131, in _determine_components_hook\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     self._encoder_config = self._get_encoder_config(\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\models\\catalog.py\", line 308, in _get_encoder_config\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     model_config_dict[\"conv_filters\"] = get_filter_config(\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\models\\utils.py\", line 226, in get_filter_config\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m     raise ValueError(\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m ValueError: No default configuration for obs shape [128, 128, 3], you must specify `conv_filters` manually as a model option. Default configurations are only available for inputs of the following shapes: [42, 42, K], [84, 84, K], [64, 64, K], [10, 10, K], [240, 320, K], and [480, 640, K]. You may alternatively want to use a custom model or preprocessor.\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\models\\catalog.py\", line 111, in __init__\u001b[32m [repeated 13x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 266 in __call__\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 65 in error_handler\u001b[32m [repeated 3x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\ops\\variables.py\", line 247 in _variable_v2_call\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=44860)\u001b[0m   File \"C:\\Users\\alaleh\\AppData\\Roaming\\Python\\Python310\\site-packages\\tensorflow\\python\\trackable\\base.py\", line 205 in _method_wrapper\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 377, in new_init\u001b[32m [repeated 6x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\core\\rl_module\\rl_module.py\", line 104, in build\u001b[32m [repeated 4x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\rllib\\algorithms\\dreamerv3\\dreamerv3_rl_module.py\", line 49, in setup\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\util\\tracing\\tracing_helper.py\", line 467, in _resume_span\n",
      "\u001b[36m(DreamerV3EnvRunner pid=12812)\u001b[0m   File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\function_manager.py\", line 726, in actor_method_executor\n",
      "\u001b[33m(raylet)\u001b[0m WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
      "\u001b[33m(raylet)\u001b[0m E0000 00:00:1708531703.264406   12536 wire_format_lite.cc:626] String field 'ray.rpc.WorkerTableData.exit_detail' contains invalid UTF-8 data when serializing a protocol buffer. Use the 'bytes' type if you intend to send raw bytes. \n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,269 E 16220 12536] (raylet.exe) logging.cc:97: Unhandled exception: class nlohmann::detail::type_error. what(): [json.exception.type_error.316] invalid UTF-8 byte at index 521: 0x8A\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:104: Stack trace: \n",
      "\u001b[33m(raylet)\u001b[0m  unknown\n",
      "\u001b[33m(raylet)\u001b[0m terminate\n",
      "\u001b[33m(raylet)\u001b[0m _NLG_Return2\n",
      "\u001b[33m(raylet)\u001b[0m _chkstk\n",
      "\u001b[33m(raylet)\u001b[0m RtlFindCharInUnicodeString\n",
      "\u001b[33m(raylet)\u001b[0m RtlRaiseException\n",
      "\u001b[33m(raylet)\u001b[0m RaiseException\n",
      "\u001b[33m(raylet)\u001b[0m CxxThrowException\n",
      "\u001b[33m(raylet)\u001b[0m BaseThreadInitThunk\n",
      "\u001b[33m(raylet)\u001b[0m RtlUserThreadStart\n",
      "\u001b[33m(raylet)\u001b[0m *** SIGABRT received at time=1708531703 ***\n",
      "\u001b[33m(raylet)\u001b[0m     @   00007FFADC31F601  (unknown)  abort\n",
      "\u001b[33m(raylet)\u001b[0m     @   00007FF6CDAD6A76  (unknown)  (unknown)\n",
      "\u001b[33m(raylet)\u001b[0m     @   00007FFADC31ED32  (unknown)  terminate\n",
      "\u001b[33m(raylet)\u001b[0m     @   00007FFABA1D1AAB  (unknown)  _NLG_Return2\n",
      "\u001b[33m(raylet)\u001b[0m     @   00007FFADEFD441F  (unknown)  _chkstk\n",
      "\u001b[33m(raylet)\u001b[0m     @   00007FFADEF4E466  (unknown)  RtlFindCharInUnicodeString\n",
      "\u001b[33m(raylet)\u001b[0m     @   00007FFADEF84475  (unknown)  RtlRaiseException\n",
      "\u001b[33m(raylet)\u001b[0m     @   00007FFADC935B0C  (unknown)  RaiseException\n",
      "\u001b[33m(raylet)\u001b[0m     @   00007FFABA1E6720  (unknown)  CxxThrowException\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361: *** SIGABRT received at time=1708531703 ***\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361:     @   00007FFADC31F601  (unknown)  abort\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361:     @   00007FF6CDAD6A76  (unknown)  (unknown)\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361:     @   00007FFADC31ED32  (unknown)  terminate\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361:     @   00007FFABA1D1AAB  (unknown)  _NLG_Return2\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361:     @   00007FFADEFD441F  (unknown)  _chkstk\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361:     @   00007FFADEF4E466  (unknown)  RtlFindCharInUnicodeString\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361:     @   00007FFADEF84475  (unknown)  RtlRaiseException\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361:     @   00007FFADC935B0C  (unknown)  RaiseException\n",
      "\u001b[33m(raylet)\u001b[0m [2024-02-22 01:08:23,280 E 16220 12536] (raylet.exe) logging.cc:361:     @   00007FFABA1E6720  (unknown)  CxxThrowException\n",
      "\u001b[36m(pid=gcs_server)\u001b[0m E0000 00:00:1708531703.265357   21708 wire_format_lite.cc:626] String field 'ray.rpc.WorkerTableData.exit_detail' contains invalid UTF-8 data when parsing a protocol buffer. Use the 'bytes' type if you intend to send raw bytes. \n",
      "2024-02-22 01:08:37,665\tERROR tune_controller.py:1374 -- Trial task failed for trial DreamerV3_my_env_d84dc_00000\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\air\\execution\\_internal\\event_manager.py\", line 110, in resolve_future\n",
      "    result = ray.get(future)\n",
      "  File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\auto_init_hook.py\", line 22, in auto_init_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\client_mode_hook.py\", line 103, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"c:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\worker.py\", line 2626, in get\n",
      "    raise value\n",
      "ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.\n",
      "\tclass_name: DreamerV3\n",
      "\tactor_id: b78ce2aabdb8bed9cbb9438701000000\n",
      "\tpid: 13244\n",
      "\tnamespace: 89cf0e42-035a-4da4-a420-4385995a93ca\n",
      "\tip: 127.0.0.1\n",
      "The actor is dead because its owner has died. Owner Id: 01000000ffffffffffffffffffffffffffffffffffffffffffffffff Owner Ip address: 127.0.0.1 Owner worker exit type: SYSTEM_ERROR Worker exit detail: Owner's node has crashed.\n",
      "The actor never ran - it was cancelled before it started running.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33m(raylet)\u001b[0m The node with node id: 782d651f52250a124dc56ef675356fe8fcba648928cd226bab51ae7a and address: 127.0.0.1 and node name: 127.0.0.1 has been marked dead because the detector has missed too many heartbeats from it. This can happen when a \t(1) raylet crashes unexpectedly (OOM, preempted node, etc.) \n",
      "\t(2) raylet has lagging heartbeats due to slow network or busy workload.\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "DreamerV3_my_env_d84dc_00000",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mRayActorError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\air\\execution\\_internal\\event_manager.py:110\u001b[0m, in \u001b[0;36mRayEventManager.resolve_future\u001b[1;34m(self, future)\u001b[0m\n\u001b[0;32m    109\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 110\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mray\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfuture\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    111\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\auto_init_hook.py:22\u001b[0m, in \u001b[0;36mwrap_auto_init.<locals>.auto_init_wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     21\u001b[0m auto_init_ray()\n\u001b[1;32m---> 22\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m fn(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\client_mode_hook.py:103\u001b[0m, in \u001b[0;36mclient_mode_hook.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    102\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(ray, func\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m)(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m--> 103\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\_private\\worker.py:2626\u001b[0m, in \u001b[0;36mget\u001b[1;34m(object_refs, timeout)\u001b[0m\n\u001b[0;32m   2625\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m-> 2626\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m value\n\u001b[0;32m   2628\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_individual_id:\n",
      "\u001b[1;31mRayActorError\u001b[0m: The actor died unexpectedly before finishing this task.\n\tclass_name: DreamerV3\n\tactor_id: b78ce2aabdb8bed9cbb9438701000000\n\tpid: 13244\n\tnamespace: 89cf0e42-035a-4da4-a420-4385995a93ca\n\tip: 127.0.0.1\nThe actor is dead because its owner has died. Owner Id: 01000000ffffffffffffffffffffffffffffffffffffffffffffffff Owner Ip address: 127.0.0.1 Owner worker exit type: SYSTEM_ERROR Worker exit detail: Owner's node has crashed.\nThe actor never ran - it was cancelled before it started running.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 43\u001b[0m\n\u001b[0;32m     31\u001b[0m config \u001b[38;5;241m=\u001b[39m DreamerV3Config()\u001b[38;5;241m.\u001b[39menvironment(env\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmy_env\u001b[39m\u001b[38;5;124m\"\u001b[39m, disable_env_checking\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, render_env\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\u001b[38;5;241m.\u001b[39mrollouts(num_rollout_workers\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m10\u001b[39m, remote_worker_envs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\u001b[38;5;241m.\u001b[39mtraining(model\u001b[38;5;241m=\u001b[39mmodel_dict)\u001b[38;5;241m.\u001b[39mresources(num_gpus\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[0;32m     33\u001b[0m tuner \u001b[38;5;241m=\u001b[39m tune\u001b[38;5;241m.\u001b[39mTuner(\n\u001b[0;32m     34\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDreamerV3\u001b[39m\u001b[38;5;124m\"\u001b[39m,  \u001b[38;5;66;03m# 2. Specify the algorithm name\u001b[39;00m\n\u001b[0;32m     35\u001b[0m     run_config\u001b[38;5;241m=\u001b[39mtrain\u001b[38;5;241m.\u001b[39mRunConfig(\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     41\u001b[0m     param_space\u001b[38;5;241m=\u001b[39mconfig,\n\u001b[0;32m     42\u001b[0m )\n\u001b[1;32m---> 43\u001b[0m results \u001b[38;5;241m=\u001b[39m \u001b[43mtuner\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     44\u001b[0m best_result \u001b[38;5;241m=\u001b[39m results\u001b[38;5;241m.\u001b[39mget_best_result(metric\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mepisode_reward_mean\u001b[39m\u001b[38;5;124m\"\u001b[39m, mode\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m     45\u001b[0m best_checkpoint \u001b[38;5;241m=\u001b[39m best_result\u001b[38;5;241m.\u001b[39mcheckpoint\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\tuner.py:381\u001b[0m, in \u001b[0;36mTuner.fit\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    379\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_ray_client:\n\u001b[0;32m    380\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 381\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_local_tuner\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    382\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m TuneError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m    383\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m TuneError(\n\u001b[0;32m    384\u001b[0m             _TUNER_FAILED_MSG\u001b[38;5;241m.\u001b[39mformat(\n\u001b[0;32m    385\u001b[0m                 path\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_local_tuner\u001b[38;5;241m.\u001b[39mget_experiment_checkpoint_dir()\n\u001b[0;32m    386\u001b[0m             )\n\u001b[0;32m    387\u001b[0m         ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\impl\\tuner_internal.py:509\u001b[0m, in \u001b[0;36mTunerInternal.fit\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    507\u001b[0m param_space \u001b[38;5;241m=\u001b[39m copy\u001b[38;5;241m.\u001b[39mdeepcopy(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mparam_space)\n\u001b[0;32m    508\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_restored:\n\u001b[1;32m--> 509\u001b[0m     analysis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit_internal\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrainable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparam_space\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    510\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    511\u001b[0m     analysis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fit_resume(trainable, param_space)\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\impl\\tuner_internal.py:628\u001b[0m, in \u001b[0;36mTunerInternal._fit_internal\u001b[1;34m(self, trainable, param_space)\u001b[0m\n\u001b[0;32m    615\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Fitting for a fresh Tuner.\"\"\"\u001b[39;00m\n\u001b[0;32m    616\u001b[0m args \u001b[38;5;241m=\u001b[39m {\n\u001b[0;32m    617\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_tune_run_arguments(trainable),\n\u001b[0;32m    618\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mdict\u001b[39m(\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    626\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_tuner_kwargs,\n\u001b[0;32m    627\u001b[0m }\n\u001b[1;32m--> 628\u001b[0m analysis \u001b[38;5;241m=\u001b[39m run(\n\u001b[0;32m    629\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39margs,\n\u001b[0;32m    630\u001b[0m )\n\u001b[0;32m    631\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclear_remote_string_queue()\n\u001b[0;32m    632\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m analysis\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\tune.py:1002\u001b[0m, in \u001b[0;36mrun\u001b[1;34m(run_or_experiment, name, metric, mode, stop, time_budget_s, config, resources_per_trial, num_samples, storage_path, storage_filesystem, search_alg, scheduler, checkpoint_config, verbose, progress_reporter, log_to_file, trial_name_creator, trial_dirname_creator, sync_config, export_formats, max_failures, fail_fast, restore, resume, reuse_actors, raise_on_failed_trial, callbacks, max_concurrent_trials, keep_checkpoints_num, checkpoint_score_attr, checkpoint_freq, checkpoint_at_end, chdir_to_trial_dir, local_dir, _remote, _remote_string_queue, _entrypoint)\u001b[0m\n\u001b[0;32m   1000\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   1001\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m runner\u001b[38;5;241m.\u001b[39mis_finished() \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m experiment_interrupted_event\u001b[38;5;241m.\u001b[39mis_set():\n\u001b[1;32m-> 1002\u001b[0m         \u001b[43mrunner\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstep\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1003\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m has_verbosity(Verbosity\u001b[38;5;241m.\u001b[39mV1_EXPERIMENT):\n\u001b[0;32m   1004\u001b[0m             _report_progress(runner, progress_reporter)\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\execution\\tune_controller.py:728\u001b[0m, in \u001b[0;36mTuneController.step\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    725\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_maybe_add_actors()\n\u001b[0;32m    727\u001b[0m \u001b[38;5;66;03m# Handle one event\u001b[39;00m\n\u001b[1;32m--> 728\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_actor_manager\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnext\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.1\u001b[39;49m\u001b[43m)\u001b[49m:\n\u001b[0;32m    729\u001b[0m     \u001b[38;5;66;03m# If there are no actors running, warn about potentially\u001b[39;00m\n\u001b[0;32m    730\u001b[0m     \u001b[38;5;66;03m# insufficient resources\u001b[39;00m\n\u001b[0;32m    731\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_actor_manager\u001b[38;5;241m.\u001b[39mnum_live_actors:\n\u001b[0;32m    732\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_insufficient_resources_manager\u001b[38;5;241m.\u001b[39mon_no_available_trials(\n\u001b[0;32m    733\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_trials()\n\u001b[0;32m    734\u001b[0m         )\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\air\\execution\\_internal\\actor_manager.py:222\u001b[0m, in \u001b[0;36mRayActorManager.next\u001b[1;34m(self, timeout)\u001b[0m\n\u001b[0;32m    219\u001b[0m [future] \u001b[38;5;241m=\u001b[39m ready\n\u001b[0;32m    221\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m future \u001b[38;5;129;01min\u001b[39;00m actor_state_futures:\n\u001b[1;32m--> 222\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_actor_state_events\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mresolve_future\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfuture\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    223\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m future \u001b[38;5;129;01min\u001b[39;00m actor_task_futures:\n\u001b[0;32m    224\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_actor_task_events\u001b[38;5;241m.\u001b[39mresolve_future(future)\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\air\\execution\\_internal\\event_manager.py:113\u001b[0m, in \u001b[0;36mRayEventManager.resolve_future\u001b[1;34m(self, future)\u001b[0m\n\u001b[0;32m    111\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m    112\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m on_error:\n\u001b[1;32m--> 113\u001b[0m         \u001b[43mon_error\u001b[49m\u001b[43m(\u001b[49m\u001b[43me\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    114\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    115\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m e\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\air\\execution\\_internal\\actor_manager.py:386\u001b[0m, in \u001b[0;36mRayActorManager._try_start_actors.<locals>.create_callbacks.<locals>.on_error\u001b[1;34m(exception)\u001b[0m\n\u001b[0;32m    385\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mon_error\u001b[39m(exception: \u001b[38;5;167;01mException\u001b[39;00m):\n\u001b[1;32m--> 386\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_actor_start_failed\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    387\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtracked_actor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtracked_actor\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexception\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexception\u001b[49m\n\u001b[0;32m    388\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\air\\execution\\_internal\\actor_manager.py:259\u001b[0m, in \u001b[0;36mRayActorManager._actor_start_failed\u001b[1;34m(self, tracked_actor, exception)\u001b[0m\n\u001b[0;32m    256\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cleanup_actor(tracked_actor\u001b[38;5;241m=\u001b[39mtracked_actor)\n\u001b[0;32m    258\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m tracked_actor\u001b[38;5;241m.\u001b[39m_on_error:\n\u001b[1;32m--> 259\u001b[0m     \u001b[43mtracked_actor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_on_error\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtracked_actor\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexception\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\execution\\tune_controller.py:1215\u001b[0m, in \u001b[0;36mTuneController._actor_failed\u001b[1;34m(self, tracked_actor, exception)\u001b[0m\n\u001b[0;32m   1209\u001b[0m     logger\u001b[38;5;241m.\u001b[39mdebug(\n\u001b[0;32m   1210\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTrial \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtrial\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m failed in its creation task. Unstaging \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   1211\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mto allow it to be re-scheduled.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   1212\u001b[0m     )\n\u001b[0;32m   1214\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_unstage_trial_with_resources(trial)\n\u001b[1;32m-> 1215\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_trial_task_failure\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrial\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexception\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexception\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1217\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_actor_manager\u001b[38;5;241m.\u001b[39mclear_actor_task_futures(tracked_actor)\n\u001b[0;32m   1219\u001b[0m \u001b[38;5;66;03m# Clean up actor\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\execution\\tune_controller.py:1375\u001b[0m, in \u001b[0;36mTuneController._trial_task_failure\u001b[1;34m(self, trial, exception)\u001b[0m\n\u001b[0;32m   1373\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_print_trial_errors:\n\u001b[0;32m   1374\u001b[0m     logger\u001b[38;5;241m.\u001b[39merror(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTrial task failed for trial \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtrial\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, exc_info\u001b[38;5;241m=\u001b[39mexception)\n\u001b[1;32m-> 1375\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_process_trial_failure\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrial\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexception\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexception\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\execution\\tune_controller.py:1401\u001b[0m, in \u001b[0;36mTuneController._process_trial_failure\u001b[1;34m(self, trial, exception)\u001b[0m\n\u001b[0;32m   1399\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_search_alg\u001b[38;5;241m.\u001b[39mon_trial_complete(trial\u001b[38;5;241m.\u001b[39mtrial_id, error\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m   1400\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_schedule_trial_stop(trial, exception\u001b[38;5;241m=\u001b[39mexception)\n\u001b[1;32m-> 1401\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_callbacks\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mon_trial_error\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1402\u001b[0m \u001b[43m    \u001b[49m\u001b[43miteration\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_iteration\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtrials\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_trials\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtrial\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrial\u001b[49m\n\u001b[0;32m   1403\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\callback.py:424\u001b[0m, in \u001b[0;36mCallbackList.on_trial_error\u001b[1;34m(self, **info)\u001b[0m\n\u001b[0;32m    422\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mon_trial_error\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39minfo):\n\u001b[0;32m    423\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m callback \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_callbacks:\n\u001b[1;32m--> 424\u001b[0m         callback\u001b[38;5;241m.\u001b[39mon_trial_error(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39minfo)\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\tune\\logger\\logger.py:167\u001b[0m, in \u001b[0;36mLoggerCallback.on_trial_error\u001b[1;34m(self, iteration, trials, trial, **info)\u001b[0m\n\u001b[0;32m    164\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mon_trial_error\u001b[39m(\n\u001b[0;32m    165\u001b[0m     \u001b[38;5;28mself\u001b[39m, iteration: \u001b[38;5;28mint\u001b[39m, trials: List[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTrial\u001b[39m\u001b[38;5;124m\"\u001b[39m], trial: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTrial\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39minfo\n\u001b[0;32m    166\u001b[0m ):\n\u001b[1;32m--> 167\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlog_trial_end\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrial\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfailed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\air\\integrations\\wandb.py:684\u001b[0m, in \u001b[0;36mWandbLoggerCallback.log_trial_end\u001b[1;34m(self, trial, failed)\u001b[0m\n\u001b[0;32m    683\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mlog_trial_end\u001b[39m(\u001b[38;5;28mself\u001b[39m, trial: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTrial\u001b[39m\u001b[38;5;124m\"\u001b[39m, failed: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[1;32m--> 684\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_signal_logging_actor_stop\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrial\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrial\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    685\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cleanup_logging_actors()\n",
      "File \u001b[1;32mc:\\Users\\alaleh\\anaconda3\\envs\\rl_env\\lib\\site-packages\\ray\\air\\integrations\\wandb.py:665\u001b[0m, in \u001b[0;36mWandbLoggerCallback._signal_logging_actor_stop\u001b[1;34m(self, trial)\u001b[0m\n\u001b[0;32m    664\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_signal_logging_actor_stop\u001b[39m(\u001b[38;5;28mself\u001b[39m, trial: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTrial\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m--> 665\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_trial_queues\u001b[49m\u001b[43m[\u001b[49m\u001b[43mtrial\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mput((_QueueItem\u001b[38;5;241m.\u001b[39mEND, \u001b[38;5;28;01mNone\u001b[39;00m))\n",
      "\u001b[1;31mKeyError\u001b[0m: DreamerV3_my_env_d84dc_00000"
     ]
    }
   ],
   "source": [
    "from ray.rllib.algorithms.dreamerv3 import DreamerV3Config\n",
    "from envs.gym_env import MyEnv\n",
    "from ray import train, tune, air\n",
    "from ray.rllib.algorithms.sac import SACConfig\n",
    "from ray.rllib.algorithms.algorithm import Algorithm\n",
    "from ray.air.integrations.wandb import WandbLoggerCallback\n",
    "\n",
    "def env_creator(env_config):\n",
    "    return MyEnv()\n",
    "\n",
    "tune.register_env(\"my_env\", env_creator)\n",
    "\n",
    "# config = (  # 1. Configure the algorithm,\n",
    "#     DreamerV3Config()\n",
    "#     .environment('my_env')\n",
    "#     .training(\n",
    "#         model_size=\"XS\", # XS, S, M, L, XL\n",
    "#     )\n",
    "# )\n",
    "\n",
    "# https://docs.ray.io/en/latest/rllib/rllib-training.html#specifying-resources\n",
    "\n",
    "model_dict = {\"conv_filters\": [\n",
    "            [16, [8, 8], 4],\n",
    "            [32, [4, 4], 4],\n",
    "            [64, [4, 4], 2],\n",
    "            [128, [4, 4], 2]\n",
    "            ]}\n",
    "\n",
    "#config = SACConfig().environment(MyEnv, disable_env_checking=True, render_env=True).rollouts(num_rollout_workers=5).training(train_batch_size=64, model=model_dict).framework(framework=\"torch\").resources(num_gpus=1)\n",
    "config = DreamerV3Config().environment(env=\"my_env\", disable_env_checking=True, render_env=True).rollouts(num_rollout_workers=10, remote_worker_envs=False).training(model=model_dict).resources(num_gpus=1)\n",
    "\n",
    "tuner = tune.Tuner(\n",
    "    \"DreamerV3\",  # 2. Specify the algorithm name\n",
    "    run_config=train.RunConfig(\n",
    "        stop={\"episode_reward_mean\": 30\n",
    "              #\"timesteps_total\": 100000,\n",
    "              },  # 学習の終了条件\n",
    "        callbacks=[WandbLoggerCallback(project=\"Wandb_nearme\")]  # コメントアウトを外すとwandbにログが送信される(https://docs.wandb.ai/quickstart)\n",
    "    ),\n",
    "    param_space=config,\n",
    ")\n",
    "results = tuner.fit()\n",
    "best_result = results.get_best_result(metric=\"episode_reward_mean\", mode=\"max\")\n",
    "best_checkpoint = best_result.checkpoint\n",
    "\n",
    "algo = Algorithm.from_checkpoint(best_checkpoint)\n",
    "\n",
    "env = MyEnv()\n",
    "episode_reward = 0\n",
    "terminated = truncated = False\n",
    "\n",
    "obs, info = env.reset()\n",
    "while not terminated and not truncated:\n",
    "    env.render()\n",
    "    action = algo.compute_single_action(obs)\n",
    "    obs, reward, terminated, truncated, info = env.step(action)\n",
    "    episode_reward += reward\n",
    "print(f\"Episode reward :{episode_reward}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rl_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
